pkg/kubelet/cm/devicemanager/manager_test.go

```diff
@@ -1403,3 +1405,78 @@ func TestReadPreNUMACheckpoint(t *testing.T) {
 	err = m.readCheckpoint()
 	require.NoError(t, err)
 }
+
+func TestGetTopologyHintsWithUpdates(t *testing.T) {
+	socketDir, socketName, _, err := tmpSocketDir()
+	defer os.RemoveAll(socketDir)
+	require.NoError(t, err)
+
+	devs := []pluginapi.Device{}
+	for i := 0; i < 1000; i++ {
+		devs = append(devs, pluginapi.Device{
+			ID:     fmt.Sprintf("dev-%d", i),
+			Health: pluginapi.Healthy,
+			Topology: &pluginapi.TopologyInfo{
+				Nodes: []*pluginapi.NUMANode{
+					{ID: 0},
+				},
+			}})
+	}
+	topology := []cadvisorapi.Node{
+		{Id: 0},
+	}
+	testCases := []struct {
+		description string
+		count       int
+		devices     []pluginapi.Device
+		testfunc    func(manager *wrappedManagerImpl)
+	}{
+		{
+			description: "getAvailableDevices data race when update device",
+			count:       1,
+			devices:     devs,
+			testfunc: func(manager *wrappedManagerImpl) {
+				manager.getAvailableDevices(testResourceName)
+			},
+		},
+		{
+			description: "generateDeviceTopologyHints data race when update device",
+			count:       1,
+			devices:     devs,
+			testfunc: func(manager *wrappedManagerImpl) {
+				manager.generateDeviceTopologyHints(
+					testResourceName, sets.NewString(), sets.NewString(), 1)
+			},
+		},
+		{
+			description: "deviceHasTopologyAlignment data race when update device",
+			count:       1000,
+			devices:     devs[:1],
+			testfunc: func(manager *wrappedManagerImpl) {
+				manager.deviceHasTopologyAlignment(testResourceName)
+			},
+		},
+	}
+
+	for _, test := range testCases {
+		t.Run(test.description, func(t *testing.T) {
+			m, _ := setupDeviceManager(t, nil, nil, socketName, topology)
+			mimpl := m.(*wrappedManagerImpl)
+
+			updated := atomic.Bool{}
+			updated.Store(false)
+			go func() {
+				for i := 0; i < test.count; i++ {
+					// simulate the device plugin to send device updates
+					mimpl.genericDeviceUpdateCallback(testResourceName, devs)
+				}
+				updated.Store(true)
+			}()
+			for !updated.Load() {
```

Comment:
What's the purpose of this condition? For me it looks like we intentionally skip calling test.testfunc call if go func() is fast enough. In this case we don't test anything, do we?

Reply:
I think this should be waiting the goroutines finish, no?

Reply:
Use the atom to identify whether the update has finished. If the update is finished, keep running the test is a waste of CPU time, and `WaitGroup` can not test whether it has been `Done()`.